idaes/core/util/scaling.py

```diff
@@ -1041,3 +1047,165 @@ def set_derivative_factor_from_state(self, deriv, nominal_wrt=1.0):
             nominal_state = 1 / scaling_factor[state_data]
             nominal_deriv = nominal_state / nominal_wrt
             scaling_factor[dv] = 1 / nominal_deriv
+
+
+# New functions
+class ScalingFactorExtractionVisitor(EXPR.StreamBasedExpressionVisitor):
+    """
+    Expression walker for collecting scaling factors in an expression.
+
+    Returns a list of scaling factors for each additive term in the expression.
+    """
+
+    def __init__(self, warning: bool = True, exception: bool = False):
+        """
+        Visitor class used to determine scaling factors of an expression. Do not use
+        this class directly.
+
+        Args:
+            warning: bool indicating whether to log a warning when a
+                missing scaling factors is encountered (default=True)
+            exception: bool indicating whether to raise an Exception when a
+                missing scaling factors is encountered (default=False)
+
+        Notes
+        -----
+        This class inherits from the :class:`StreamBasedExpressionVisitor` to implement
+        a walker that returns the pyomo units and pint units corresponding to an
+        expression.
+        There are class attributes (dicts) that map the expression node type to the
+        particular method that should be called to return the units of the node based
+        on the units of its child arguments. This map is used in exitNode.
+        """
+        super().__init__()
+
+        self.warning = warning
+        self.exception = exception
+
+    def _get_scale_factor_for_sum(self, node, child_scale_factors):
+        sf = []
+        for i in child_scale_factors:
+            for j in i:
+                sf.append(j)
+        return sf
```

Comment:
For a sum of _strictly positive numbers_, a good estimate of the scaling factor of the sum is the harmonic sum of the scaling factors of the numbers (like resistors in parallel). For example, let, a ~ O(1), b ~ O(10), and c ~O(0.1). We have scaling factors s_a = 1, s_b = 0.1, and s_c = 10. The harmonic sum of these scaling factors is 1/(1/1 + 1/0.1 + 1/10) ~= 0.09. It's dominated by the smallest scaling factor (s_b) because the actual sum is dominated by the largest quantity (b). 

Say you then have a ~ O(1), b ~ O(1), and c ~ O(1), and so s_a = 1, s_b =1 , and s_c =1. You'd then have for a scaling factor for the sum 1/(1+1+1) = 1/3.

I'm not sure what exact structure `child_scaling_factor` is supposed to have, nor what structure the `sf` list is supposed to have, so I cannot evaluate this code on that.

Reply:
The idea here was just to get the scaling factors for each term - we would then decide on what to do with them later. The aim was to let us support a range of different approaches (maximum, minimum, harmonic sum, etc.).